Adds proxy support for HTTPClient

timothyqiu opened the pull request:
This PR adds HTTP(s) proxy support for `HTTPClient`.

* `core/io/http_client.{h,cpp}`:
    * Added `set_http{,s}_proxy(host, port)` methods to set/unset proxy. They are not pure virtual functions, the default implementation does nothing and emits warnings about proxy not implemented.
* `core/io/http_client_tcp.{h,cpp}`:
    * what HTTP proxy does:
        1. connect to proxy server instead of target server
        2. use full URL in request line
    * what HTTPS proxy does:
        1. connect to proxy server instead of target server
        2. send a `CONNECT` request to setup the tunnel to target server
        3. do SSL handshakes and other things as usual
* `doc/classes/HTTPClient.xml`: added corresponding doc for new methods

Some notes:

* Common HTTP proxy servers can send requests to different target servers. This implementation always closes and reconnects to the proxy server for each `connect_to_host()` since simple HTTP connections are seldom used nowadays and reusing the connection complicates the code.
* Only proxy servers that uses HTTP protocol are supported (i.e. no HTTP-on-HTTPS or HTTPS-on-HTTPS). It's hard to find such proxy servers.
* Basic authentication is not currently implemented.

<details>
<summary>Here is a messy test code.</summary>

```gdscript
extends SceneTree

func _init():
    var err = 0
    var http = HTTPClient.new() # Create the Client.

    http.set_http_proxy("localhost", 8123)
    http.set_https_proxy("127.0.0.1", 8123)

    print("----- http with proxy -----")
    err = http.connect_to_host("httpbin.org", -1, false) # Connect to host/port.
    assert(err == OK) # Make sure connection was OK.
    await do_http(http)
    http.close()

    print("----- https with proxy -----")
    err = http.connect_to_host("httpbin.org", -1, true) # Connect to host/port.
    assert(err == OK) # Make sure connection was OK.
    await do_http(http)
    http.close()

    http.set_http_proxy("127.0.0.1", -1)
    http.set_https_proxy("127.0.0.1", -1)

    print("----- http without proxy -----")
    err = http.connect_to_host("httpbin.org", -1, false) # Connect to host/port.
    assert(err == OK) # Make sure connection was OK.
    await do_http(http)
    http.close()

    print("----- https without proxy -----")
    err = http.connect_to_host("httpbin.org", -1, true) # Connect to host/port.
    assert(err == OK) # Make sure connection was OK.
    await do_http(http)
    http.close()

    quit()


func do_http(http):
    var err = 0
    # Wait until resolved and connected.
    while http.get_status() == HTTPClient.STATUS_CONNECTING or http.get_status() == HTTPClient.STATUS_RESOLVING:
        http.poll()
        print("Connecting...")
        if not OS.has_feature("web"):
            OS.delay_msec(500)
        else:
            await (Engine.get_main_loop().idle_frame)

    assert(http.get_status() == HTTPClient.STATUS_CONNECTED) # Could not connect

    # Some headers
    var headers = [
        "User-Agent: Pirulo/1.0 (Godot)",
        "Accept: */*"
    ]

    err = http.request(HTTPClient.METHOD_GET, "/get", headers) # Request a page from the site (this one was chunked..)
    assert(err == OK) # Make sure all is OK.

    while http.get_status() == HTTPClient.STATUS_REQUESTING:
        # Keep polling for as long as the request is being processed.
        http.poll()
        print("Requesting...")
        if not OS.has_feature("web"):
            OS.delay_msec(500)
        else:
            # Synchronous HTTP requests are not supported on the web,
            # so wait for the next main loop iteration.
            await (Engine.get_main_loop().idle_frame)

    assert(http.get_status() == HTTPClient.STATUS_BODY or http.get_status() == HTTPClient.STATUS_CONNECTED) # Make sure request finished well.

    print("response? ", http.has_response()) # Site might not have a response.

    if http.has_response():
        # If there is a response...

        headers = http.get_response_headers_as_dictionary() # Get response headers.
        print("code: ", http.get_response_code()) # Show response code.
        print("**headers:\\n", headers) # Show headers.

        # Getting the HTTP Body

        if http.is_response_chunked():
            # Does it use chunks?
            print("Response is Chunked!")
        else:
            # Or just plain Content-Length
            var bl = http.get_response_body_length()
            print("Response Length: ", bl)

        # This method works for both anyway

        var rb = PackedByteArray() # Array that will hold the data.

        while http.get_status() == HTTPClient.STATUS_BODY:
            # Get a chunk.
            var chunk = http.read_response_body_chunk()
            if chunk.size() == 0:
                if not OS.has_feature("web"):
                    # Got nothing, wait for buffers to fill a bit.
                    OS.delay_usec(1000)
                else:
                    await (Engine.get_main_loop().idle_frame)
            else:
                rb = rb + chunk # Append to read buffer.
            # While there is body left to be read
            http.poll()
        # Done!

        print("bytes got: ", rb.size())
        var text = rb.get_string_from_ascii()
        print("Text: ", text)
```

</details>

Calinou commented:
Does this PR make use of the `HTTP_PROXY` and `HTTPS_PROXY` environment variables? If so, that would resolve https://github.com/godotengine/godot/issues/6100 :slightly_smiling_face: 

(Also, since environment variables are case-sensitive on macOS and Linux, I'd try to read both `HTTP_PROXY` then `http_proxy`.)

timothyqiu commented:
No. This PR only adds an API to setup proxy servers manually. No proxy is used by default.

And I'm indeed heading toward adding proxy for AssetLib and detecting environment variables. This PR and #43493 are preparations :)

Faless commented:
I wish this had gone through a proposal first. I'm not even sure if we want to support http proxies out of the box, but in any case, this overcomplicates the HTTPClient state machine.
Ideally, if you set a proxy, it should be transparent from the state machine point of view (no new states at all). But then again, I'm not even sure if we want to implement this as it should be implementable in GDScript.
`HTTPClient` is already hard to maintain as it is, adding more optional features like this without some refactoring is going to make it worse.

timothyqiu commented:
Although seems like an enhancement, this was an attempt to eventually fix #6100, so no proposal was made. AssetLib is not working behind firewalls (it's basically unusable in China due to the Great FireWall).

I think I can change the way of setting up proxy server if the requirement is to leave the state machine untouched. (But the user won't be able to do proper error handling then.)

timothyqiu commented:
Updated to leave the state machine alone. Proxy error now uses `STATUS_CANT_CONNECT`.

timothyqiu committed:
Adds proxy support for HTTPClient

Also fixed a potential infinite loop when connecting to server.

timothyqiu commented:
Rebased and moved most codes into to the recently refactored `HTTPClientTCP` class.

akien-mga commented:
We just discussed this in a PR review meeting and @faless and @mhilbrunner agreed that this can be merged. Needs a rebase, and Max plans to do some final testing to validate.

timothyqiu commented:
Rebased.

Also fixed a potential infinite loop when connecting to server:
https://github.com/godotengine/godot/blob/efd0db85f6fef4414aa6eb839a72ca75c2feb54b/core/io/http_client_tcp.cpp#L267-L272

The `front()` here should be `pop_front()`. Since this PR also needs to change `conn_port` to `server_port`, I added the typo-fix into this PR for easy testing.

mhilbrunner reviewed (approved):
I finally found time to do a more in depth review and testing, using your provided sample code as well as my own with a company proxy server and also [mitmproxy](https://mitmproxy.org/).

I found no issues, and everything seems to work. Good work! And I apologize that this took so very long to review.

mhilbrunner commented:
@akien-mga @Faless I think we can merge this now, not sure whether we want to cherry-pick - to fix the Asset Library we need to add code to actually use this new feature.

akien-mga commented:
Thanks!

> And I'm somewhat afraid of doing major networking changes in minor releases :P

Well `3.x` releases are "minor" only in semver terminology, in terms of actual features and API changes they're fairly major. Moreover this PR doesn't seem to do major changes to the current API, everything added is optional, no? So I think it would be good to cherry-pick for 3.5. And indeed a follow-up PR to fix using the Asset Library behind a proxy would be great :+1:
